File: compiler/extensions/cpp/runtime/test/zserio/JsonParserTest.cpp
#include "gtest/gtest.h"

#include <sstream>
#include "zserio/JsonParser.h"

namespace zserio
{

namespace
{

class DummyObserver : public JsonParser::IObserver
{
public:
    const std::vector<std::string>& getReport() const
    {
        return m_report;
    }

    virtual void beginObject() override
    {
        m_report.push_back("beginObject");
    }

    virtual void endObject() override
    {
        m_report.push_back("endObject");
    }

    virtual void beginArray() override
    {
        m_report.push_back("beginArray");
    }

    virtual void endArray() override
    {
        m_report.push_back("endArray");
    }

    virtual void visitKey(StringView stringValue) override
    {
        m_report.push_back("visitKey: " + toString(stringValue));
    }

    virtual void visitValue(std::nullptr_t) override
    {
        m_report.push_back("visitValue: null");
    }

    virtual void visitValue(bool boolValue) override
    {
        m_report.push_back("visitValue: " + toString(boolValue));
    }

    virtual void visitValue(int64_t intValue) override
    {
        m_report.push_back("visitValue: " + toString(intValue));
    }

    virtual void visitValue(uint64_t uintValue) override
    {
        m_report.push_back("visitValue: " + toString(uintValue));
    }

    virtual void visitValue(double doubleValue) override
    {
        m_report.push_back("visitValue: " + std::to_string(doubleValue));
    }

    virtual void visitValue(StringView stringValue) override
    {
        m_report.push_back("visitValue: " + toString(stringValue));
    }

private:
    std::vector<std::string> m_report;
};

void assertStartsWith(const std::string& expectedStart, const char* cstr)
{
    std::string str(cstr);
    if (str.size() > expectedStart.size())
        str.resize(expectedStart.size());
    ASSERT_EQ(expectedStart, str);
}

} // namespace

TEST(JsonParserTest, empty)
{
    std::stringstream str("");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_TRUE(jsonParser.parse());
    ASSERT_EQ(0, observer.getReport().size());
}

TEST(JsonParserTest, oneString)
{
    std::stringstream str("\"text\"");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_TRUE(jsonParser.parse());
    ASSERT_EQ(std::vector<std::string>{{"visitValue: text"}}, observer.getReport());
}

TEST(JsonParserTest, twoStrings)
{
    std::stringstream str("\"text\"\"second\"");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_FALSE(jsonParser.parse());
    std::vector<std::string> expectedReport = {{"visitValue: text"}};
    ASSERT_EQ(expectedReport, observer.getReport());
    ASSERT_TRUE(jsonParser.parse());
    expectedReport.push_back("visitValue: second");
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, parse)
{
    std::stringstream str("{\"array\":\n[\n{\"key1\":\n10, \"key2\":\n\"text\"}, {}]}");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_TRUE(jsonParser.parse());

    std::vector<std::string> expectedReport = {{
        {"beginObject"},
        {"visitKey: array"},
        {"beginArray"},
        {"beginObject"},
        {"visitKey: key1"},
        {"visitValue: 10"},
        {"visitKey: key2"},
        {"visitValue: text"},
        {"endObject"},
        {"beginObject"},
        {"endObject"},
        {"endArray"},
        {"endObject"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, valueTypes)
{
    std::stringstream str("null true 10 -10 1.1 \"str\"");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    while (!jsonParser.parse())
    {}

    std::vector<std::string> expectedReport = {{
        {"visitValue: null"},
        {"visitValue: true"},
        {"visitValue: 10"},
        {"visitValue: -10"},
        {"visitValue: 1.100000"},
        {"visitValue: str"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, unexpectedObject)
{
    std::stringstream str("{\n\n{\n\n");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 3:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        "beginObject"
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, unexpectedObjectAfterItemSeparator)
{
    std::stringstream str("{\n  \"key\": 10,\n  {\n");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 3:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        {"beginObject"},
        {"visitKey: key"},
        {"visitValue: 10"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, missingObjectItemSeparator)
{
    std::stringstream str("{\n\"item1\":\"text\"\n\"item2\":\"text\"\n}");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 3:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        {"beginObject"},
        {"visitKey: item1"},
        {"visitValue: text"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, wrongKeyType)
{
    std::stringstream str("{\n10:\"text\"\n}");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 2:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        "beginObject"
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, unexpectedElementToken)
{
    std::stringstream str("{\n\"item\":}");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 2:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        {"beginObject"},
        {"visitKey: item"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, missingArrayElementSeparator)
{
    std::stringstream str("{\n\"array\":\n[10\n20\n]}");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 4:", e.what());
            throw;
        }
    }, JsonParserException);

    std::vector<std::string> expectedReport = {{
        {"beginObject"},
        {"visitKey: array"},
        {"beginArray"},
        {"visitValue: 10"}
    }};
    ASSERT_EQ(expectedReport, observer.getReport());
}

TEST(JsonParserTest, unknownToken)
{
    std::stringstream str("\\\n");
    DummyObserver observer;
    JsonParser jsonParser(str, observer);
    ASSERT_THROW({
        try
        {
            jsonParser.parse();
        }
        catch (const JsonParserException& e)
        {
            assertStartsWith("JsonParser line 1:", e.what());
            throw;
        }
    }, JsonParserException);

    ASSERT_TRUE(observer.getReport().empty());
}

} // namespace zserio

File: compiler/extensions/java/runtime/test/zserio/runtime/json/JsonParserTest.java
package zserio.runtime.json;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.Test;

public class JsonParserTest
{
    @Test
    public void empty() throws IOException
    {
        final Reader reader = new StringReader("");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        assertTrue(jsonParser.parse());
        assertEquals(0, observer.getReport().size());

        reader.close();
    }

    @Test
    public void oneString() throws IOException
    {
        final Reader reader = new StringReader("\"text\"");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        assertTrue(jsonParser.parse());
        assertEquals(1, observer.getReport().size());
        assertEquals("visitValue: text", observer.getReport().get(0));

        reader.close();
    }

    @Test
    public void twoStrings() throws IOException
    {
        final Reader reader = new StringReader("\"text\"\"second\"");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        assertFalse(jsonParser.parse());
        assertEquals(1, observer.getReport().size());
        assertEquals("visitValue: text", observer.getReport().get(0));
        assertEquals(1, jsonParser.getLine());
        assertEquals(7, jsonParser.getColumn());

        assertTrue(jsonParser.parse());
        assertEquals(2, observer.getReport().size());
        assertEquals("visitValue: second", observer.getReport().get(1));
        assertEquals(1, jsonParser.getLine());
        assertEquals(15, jsonParser.getColumn());

        reader.close();
    }

    @Test
    public void parse() throws IOException
    {
        final Reader reader = new StringReader("{\"array\":\n[\n{\"key1\":\n10, \"key2\":\n\"text\"}, {}]}");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        assertTrue(jsonParser.parse());
        assertEquals(13, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));
        assertEquals("visitKey: array", observer.getReport().get(1));
        assertEquals("beginArray", observer.getReport().get(2));
        assertEquals("beginObject", observer.getReport().get(3));
        assertEquals("visitKey: key1", observer.getReport().get(4));
        assertEquals("visitValue: 10", observer.getReport().get(5));
        assertEquals("visitKey: key2", observer.getReport().get(6));
        assertEquals("visitValue: text", observer.getReport().get(7));
        assertEquals("endObject", observer.getReport().get(8));
        assertEquals("beginObject", observer.getReport().get(9));
        assertEquals("endObject", observer.getReport().get(10));
        assertEquals("endArray", observer.getReport().get(11));
        assertEquals("endObject", observer.getReport().get(12));

        reader.close();
    }

    @Test
    public void unexpectedObject() throws IOException
    {
        final Reader reader = new StringReader("{\n\n{\n\n");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals("JsonParser:3:1: Unexpected token: BEGIN_OBJECT ('{'), expecting END_OBJECT!",
                exception.getMessage());
        assertEquals(1, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));

        reader.close();
    }

    @Test
    public void unexpectedObjectAfterItemSeparator() throws IOException
    {
        final Reader reader = new StringReader("{\n  \"key\": 10,\n  {\n");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals("JsonParser:3:3: Unexpected token: BEGIN_OBJECT ('{'), expecting VALUE!",
                exception.getMessage());
        assertEquals(3, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));
        assertEquals("visitKey: key", observer.getReport().get(1));
        assertEquals("visitValue: 10", observer.getReport().get(2));

        reader.close();
    }

    @Test
    public void missingObjectItemSeparator() throws IOException
    {
        final Reader reader = new StringReader("{\n\"item1\":\"text\"\n\"item2\":\"text\"\n}");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals("JsonParser:3:1: Unexpected token: VALUE ('item2'), expecting END_OBJECT!",
                exception.getMessage());
        assertEquals(3, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));
        assertEquals("visitKey: item1", observer.getReport().get(1));
        assertEquals("visitValue: text", observer.getReport().get(2));

        reader.close();
    }

    @Test
    public void wrongKeyType() throws IOException
    {
        final Reader reader = new StringReader("{\n10:\"text\"\n}");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals("JsonParser:2:1: Key must be a string value!", exception.getMessage());
        assertEquals(1, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));

        reader.close();
    }

    @Test
    public void unexpectedElementToken() throws IOException
    {
        final Reader reader = new StringReader("{\n\"item\":}");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals("JsonParser:2:8: Unexpected token: END_OBJECT ('}'), expecting one of "
                        + "[BEGIN_OBJECT, BEGIN_ARRAY, VALUE]!",
                exception.getMessage());
        assertEquals(2, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));
        assertEquals("visitKey: item", observer.getReport().get(1));

        reader.close();
    }

    @Test
    public void missingArrayElementSeparator() throws IOException
    {
        final Reader reader = new StringReader("{\n\"array\":\n[10\n20\n]}");
        final DummyObserver observer = new DummyObserver();
        final JsonParser jsonParser = new JsonParser(reader, observer);

        final JsonParserError exception = assertThrows(JsonParserError.class, () -> jsonParser.parse());
        assertEquals(
                "JsonParser:4:1: Unexpected token: VALUE ('20'), expecting END_ARRAY!", exception.getMessage());
        assertEquals(4, observer.getReport().size());
        assertEquals("beginObject", observer.getReport().get(0));
        assertEquals("visitKey: array", observer.getReport().get(1));
        assertEquals("beginArray", observer.getReport().get(2));
        assertEquals("visitValue: 10", observer.getReport().get(3));

        reader.close();
    }

    private static class DummyObserver implements JsonParser.Observer
    {
        public List<String> getReport()
        {
            return report;
        }

        @Override
        public void beginObject()
        {
            report.add("beginObject");
        }

        @Override
        public void endObject()
        {
            report.add("endObject");
        }

        @Override
        public void beginArray()
        {
            report.add("beginArray");
        }

        @Override
        public void endArray()
        {
            report.add("endArray");
        }

        @Override
        public void visitKey(String key)
        {
            report.add("visitKey: " + key);
        }

        @Override
        public void visitValue(Object value)
        {
            report.add("visitValue: " + value.toString());
        }

        private final List<String> report = new ArrayList<String>();
    };
}
